Add Limit Trie Logs Early Access Feature Guide (#1495)
hyperledger/besu-docs: docs/public-networks/how-to/bonsai-limit-trie-logs.md

---
title: Limit Trie Logs for Bonsai
sidebar_position: 12

> description: Enable this early access feature to reduce the size of your database

Reviewer: Enable this feature**

tags:
  - public networks
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> # Limit Trie Logs for Bonsai

Reviewer: Bonsai... to reduce database size.

change this for SEO

Author: Are you suggesting, "Limit Trie Logs to Reduce Database Size"?

Are you happy to exclude Bonsai from the title? I wanted to make it clear it was irrelevant for FOREST, but if we plan to remove FOREST soon, maybe that doesn't matter.

## Step by Step Guide

:::caution

> We strongly recommend reading the rest of this documentation before running these commands because your node’s configuration may require changes to these commands.

Reviewer 2: ```suggestion
Limiting trie logs is an early access feature. Before executing the commands in this guide, ensure you review your node configuration.
```

:::

> *Targeting 24.1.2 release, but currently you must be on the Besu `main` branch for this to work*

Reviewer: I think we should wait on this page until the feature is released.

Author: removed this line

1. Update Besu config to add --Xbonsai-limit-trie-logs-enabled but don’t restart yet
1. Stop Besu

> 1. (optional) Run: 

Reviewer: should we provide relative paths?

Author: In this example, I opted to use the most widely used paths, which fortunately are the same for ethstaker docs and the two most popular step by step guides: Somer Esats and CoinCashew.

I put a warning to not copy and paste, but if ppl do then it should work out of the box if they followed the popular guides.

`sudo /usr/local/bin/besu/bin/besu --data-storage-format=BONSAI --data-path=/var/lib/besu --sync-mode=X_SNAP storage x-trie-log prune`
1. Start Besu (remembering to run sudo systemctl daemon-reload if you use a systemd service file as per CoinCashew and Somer)
1. Look out for `Limit trie logs enabled: retention: 512; prune window: 30000` in your Besu config printout during startup

> 1. Enjoy more GBs

Reviewer: enjoy more free space

## What?

> We have a new experimental feature available: `--Xbonsai-limit-trie-logs-enabled` which aims to keep Besu’s database as small as it can be. After a brief grace period, we intend to make this enabled by default for stakers. This is only relevant if you're using `data-storage-format=BONSAI`.

Reviewer 2: For tech docs, we typically don't write with a marketing tone and usually opt for language that is a but more succinct and to the point. I think a rewrite for this would be something like:

```suggestion
The feature `--Xbonsai-limit-trie-logs-enabled` minimizes the database size of Besu and requires the use of `data-storage-format=BONSAI`. When enabled, this feature can reduce database growth by more than 3 GB each week.
```

From our testing, we estimate this will **save users > 3 GB per week** in database growth. Early testing indicates Besu’s **overall database growth with this enabled is ~7 GB per week** (thanks Yorick!) which is on par with geth.

## Why?
Some users noticed that resyncing Besu can free up disk space, especially for longer running nodes. This is despite `--data-storage-format=BONSAI` having “implicit pruning”. More on this reddit thread: https://www.reddit.com/r/ethstaker/comments/12xnxxi/clearing_up_besubonsai_confusion_on_state_growth/

> In reality, whilst the Merkle Patricia Trie is implicitly pruned, the BONSAI feature did introduce an extra data structure: the Trie Log (more detail about BONSAI and trie logs here: https://consensys.io/blog/bonsai-tries-guide)

Reviewer 2: I am not sure this is needed in this guide

Author: Agree, this is better for the initial announcement

> In reality, whilst the Merkle Patricia Trie is implicitly pruned, the BONSAI feature did introduce an extra data structure: the Trie Log (more detail about BONSAI and trie logs here: https://consensys.io/blog/bonsai-tries-guide)

Reviewer 2: Can be removed

The Trie Logs are retained in order to cope with chain reorgs. After each block is finalized, Trie Logs older than that are no longer required so it is safe to remove them from both the node's and the network's point of view.

## How?
If you want to use this feature before it is enabled by default, simply add this option to your Besu command: `--Xbonsai-limit-trie-logs-enabled`

> When you restart Besu it will begin pruning, block by block (and a cheeky bit during Besu startup).

Reviewer: block by block after an initial reduction in the database.

> If you want **maximum database size reduction**, read on.

Reviewer: delete

If you have a long-running node, this will not immediately clear your backlog of trie logs in the same way that resyncing does. In order to do this we’re providing a “run once” **offline** command to immediately **prune all your old trie logs in a few minutes or less**. Note, this requires Besu to be shutdown before running, but downtime will be minimal. You will **not** need to run this command a second time if you keep `--Xbonsai-limit-trie-logs-enabled`.

> ## I’m impatient, reduce my database size now!

Reviewer: Remove this header

> Okay, okay - we got you! We built a one-off Besu command to remove this extra data (usually in seconds) and avoid having to resync. For minimal downtime, we recommend running this command **before** restarting Besu `--Xbonsai-limit-trie-logs-enabled` (easiest to do it all at the same time).

Reviewer: Remove okay we got you

2024-01-31 05:12:45.916+00:00 | main | INFO  | TrieLogHelper | Obtaining trielogs from db, this may take a few minutes...
2024-01-31 05:12:46.036+00:00 | main | INFO  | TrieLogHelper | Clear trie logs...
2024-01-31 05:12:46.100+00:00 | main | INFO  | TrieLogHelper | Restoring trie logs retained from batch 1...
2024-01-31 05:12:46.168+00:00 | main | INFO  | TrieLogHelper | Key(0): 0xc0a1f26bf828ae6d9af8fda8af3126f12a6d986afea05fdc61b25d0d0ab3b8d4
...
2024-01-31 05:12:46.263+00:00 | main | INFO  | TrieLogHelper | Key(511): 0x5eb700164f4190b8806c7ce3de2f6e5b4f4c05a61bba4058f1b0a5ef63ec73af
2024-01-31 05:12:46.359+00:00 | main | INFO  | TrieLogHelper | Deleting files...
2024-01-31 05:12:46.363+00:00 | main | INFO  | TrieLogSubCommand | Finished pruning - recalculating size...
2024-01-31 05:12:51.658+00:00 | main | INFO  | TrieLogSubCommand | Trie logs size after: 0 B
Prune ran successfully. We estimate you freed up 8 GiB!
```

If you use a toml config file, then you can simply do something like:

`sudo /usr/local/bin/besu/bin/besu --config-file=besu-config.toml storage x-trie-log prune`

> ## Subcommand Troubleshooting

Reviewer 2: ```suggestion
## Troubleshoot
```
